### DISCUSSION TITLE
[Question] reading compiler from env when cross-compiling

### DISCUSSION THREAD

**Discussion by Corallo** [mesonbuild/meson#11557] [Q&A] [answered]
Hello,

I am trying to cross compile a project for arm
In my system, I have an env variable $CC that contains the compiler binary and his options (generated by poky).
Something like: 
aarch64-linux-gnu-gcc -mcpu=cortex-a53 -march=armv8-a+crc+crypto -fstack-protector-strong -O2 -D_FORTIFY_SOURCE=2 -Wformat -Wformat-security -Werror=format-security --sysroot=...

I'd the cross compile file to take the compiler from this variable, something like:
[binaries]
c = $CC

Is it possible?

**Comment by xclaesse** [comment, marked as answer]
There is https://mesonbuild.com/Commands.html#env2mfile to generate the cross file from environment variables.

**Reply by Corallo** [reply]
Looks great! I'll try it, thanks!

**Reply by Corallo** [reply]
To the file generate by env2mfile, I would need to add the line
pkgconfig = 'pkg-config'
because for some reason meson doesn't find it automatically in my system, (even if it is un /usr/bin, wich is in PATH).
Do you know if there is a preferable way to add it to the file? Or if I just sed it?

**Reply by eli-schwartz** [reply]
Shouldn't this be handled by setting it from the environment as $PKG_CONFIG, since the entire idea of env2mfile is that it translates *environment variables* to a cross file?

**Reply by Corallo** [reply]
env2mfile doesn't translate PKG_CONFIG in the environment in the cross file

**Reply by Corallo** [reply]
Oh but I see that meson takes $PKG_CONFIG from env when doing setup, without the need of being in the cross configuration file. That's great!

**Reply by eli-schwartz** [reply]
Meh... I looked at the source, and for cross setup it only runs detect_compilers_from_envvars() and not detect_binaries_from_envvars().

Seems like a bug.

**Reply by Corallo** [reply]
I tried to add it, it seems to produce a more complete corss compile file.
Even before adding, it was finding C and CPP compilers, but now I get also "other binaries"


```
# Other binaries
ar = ['aarch64-linux-gnu-ar']
as = ['aarch64-linux-gnu-as']
ld = ['aarch64-linux-gnu-ld']
nm = ['aarch64-linux-gnu-nm']
objcopy = ['aarch64-linux-gnu-objcopy']
objdump = ['aarch64-linux-gnu-objdump']
pkg-config = ['pkg-config']
pkgconfig = ['pkg-config']
ranlib = ['aarch64-linux-gnu-ranlib']
readelf = ['aarch64-linux-gnu-readelf']
strip = ['aarch64-linux-gnu-strip']

```
Should I make a PR?

**Reply by xclaesse** [reply]
This should help: https://github.com/mesonbuild/meson/pull/11561
